### ISSUE TITLE
Moor should handle nulls before they get to TypeConverter

### ISSUE THREAD

**Issue by North101** [simolus3/drift#1241]
I think Moor should handle nulls before they get to TypeConverter.

```
enum DeckGroup {
  Name,
  Cycle,
  Pack,
  Side,
  Faction,
  Created,
  Updated,
  Synced,
}
```

```
class DeckGroupConverter extends TypeConverter<DeckGroup, String> {
  const DeckGroupConverter();

  @override
  DeckGroup? mapToDart(String? value) {
    if (value == null) return null;

    return DeckGroup.values.firstWhereOrNull((e) => e.toString() == value);
  }

  @override
  String? mapToSql(DeckGroup? value) {
    if (value == null) return null;

    return value.toString();
  }
}
```

```
CREATE TABLE settings(
  deck_group TEXT MAPPED BY `const DeckGroupConverter()` NOT NULL
);
```

Take the above example When I try to access settings.deckGroup (which is not nullable) its actually possible to get a null. Obviously I can easily fix this by providing a default, but I think a better solution is to to never pass nulls to TypeConverter e.g.

```
class DeckGroupConverter extends TypeConverter<DeckGroup, String> {
  const DeckGroupConverter();

  @override
  DeckGroup mapToDart(String value) {
    return DeckGroup.values.firstWhereOrNull((e) => e.toString() == value) ?? DeckGroup.default;
  }

  @override
  String mapToSql(DeckGroup value) {
    return value.toString();
  }
}
```

**Comment by westito** [comment]
+1
I also thoght about it

**Comment by simolus3** [comment]
I agree too! The current design is mainly a leftover from pre-nnbd to avoid making it too breaking. Unfortunately that also means that we need a moor 5 for this.

In the generator it will be a bit more complicated to handle when we have converters mapping non-nullable values to nullable values (or vice versa, or if both are nullable to begin with). The behavior here isn't obvious, I'll think about it some more. The non-nullable to non-nullable case, which is probably the most common, should be easy to support though.

**Comment by simolus3** [comment]
I've fixed this in 06af930941407e16b3a0f2f1bdeb5b7d9038fcb2. Essentially, the change is to always enable `null_aware_type_converters`. Together with `NullAwareTypeConverter`, I think this is quite neat to use:

- I've changed the definitions of `mapToDart` and `mapToSql` to take the types that the converter was declared with. So if you have a `TypeConverter<MyObject, String>`, it can't deal with `null`s in either direction. `drift_dev` now emits a warning when such converter is used on a nullable column. So most `NullAwareTypeConverter`s can now be regular type converters if they're only used on non-null columns.
- I've added `NullAwareTypeConverter.wrap` to wrap a non-nullable type converter into a nullable converter. This works by just mapping `null` to `NULL` and can be interesting for type converters which are sometimes applied to nullable columns and sometimes not.

Overall, I hope that these changes make it easier to reason about how `null` values interact with type converters. There may be some cases that require a bit more boilerplate with `NullAwareTypeConverter.wrap` now, but IMO the consistency is worth it. If there's a common case that I've overlooked that turns out to be much more annoying to write now, I'll take a look at adding better API shortcuts for it.
